src/equations/hyperbolic_serre_green_naghdi_1d.jl

```diff
@@ -0,0 +1,328 @@
+(bathymetry) ``b = \eta_0 - D``. The water height above the bathymetry
+is therefore given by ``h = \eta - \eta_0 + D``.
+The total water height is therefore given by ``\eta = h + b``.
+
+Two types of variable `bathymetry_type` are supported:
+- [`bathymetry_flat`](@ref): flat bathymetry (typically ``b = 0`` everywhere)
+- [`bathymetry_mild_slope`](@ref): variable bathymetry with mild-slope approximation
+
+For the mild-slope approximation, the Serre-Green-Naghdi equations are
+```math
+\begin{aligned}
+  h_t + (h v)_x &= 0,\\
+  h v_t + \frac{1}{2} g (h^2)_x + \frac{1}{2} h (v^2)_x
+    + \biggl( \frac{\lambda}{3} \eta (1 - \eta / h) \biggr)_x
+    + \biggl( g h + \frac{\lambda}{2} (1 - \eta / h) \biggr) b_x &= 0,\\
+  h w_t + h v w_x &= \lambda (1 - \eta / h),\\
+  \eta_t + \eta_x u + \frac{3}{2} b_x v &= w.
+\end{aligned}
+```
+
+References for the hyperbolized Serre-Green-Naghdi system can be found in
+- Favrie and Gavrilyuk.
+  A rapid numerical method for solving Serre-Green-Naghdi equations
+  describing long free surface gravity waves
+  [DOI: 10.1088/1361-6544/aa712d](https://doi.org/10.1088/1361-6544/aa712d)
+- Busto, Dumbser, Escalante, Favrie, and Gavrilyuk.
+  On High Order ADER Discontinuous Galerkin Schemes for First Order Hyperbolic
+  Reformulations of Nonlinear Dispersive Systems
+  [DOI: 10.1007/s10915-021-01429-8](https://doi.org/10.1007/s10915-021-01429-8)
+
+The semidiscretization implemented here conserves
+- the total water mass (integral of h) as a linear invariant
+- the total modified energy
+
+for periodic boundary conditions (see Ranocha and Ricchiuto (2024)).
+Additionally, it is well-balanced for the lake-at-rest stationary solution, see
+- Hendrik Ranocha and Mario Ricchiuto (2024)
+  Structure-preserving approximations of the Serre-Green-Naghdi
+  equations in standard and hyperbolic form
+  [arXiv: 2408.02665](https://arxiv.org/abs/2408.02665)
+"""
+struct HyperbolicSerreGreenNaghdiEquations1D{Bathymetry <: Union{BathymetryFlat, BathymetryMildSlope}, RealT <: Real} <:
+       AbstractSerreGreenNaghdiEquations{1, 5}
+    bathymetry_type::Bathymetry # type of bathymetry
+    gravity::RealT # gravitational constant
+    eta0::RealT # constant still-water surface
+    λ::RealT # hyperbolic relaxation parameter (→ ∞ for Serre-Green-Naghdi)
+end
+
+function HyperbolicSerreGreenNaghdiEquations1D(bathymetry_type = bathymetry_mild_slope;
+                                               gravity_constant,
+                                               eta0 = 0.0)
+    HyperbolicSerreGreenNaghdiEquations1D(bathymetry_type, gravity_constant, eta0)
+end
+
+# TODO: How shall we handle eta=h+b vs the new variable η?
+function varnames(::typeof(prim2prim), ::HyperbolicSerreGreenNaghdiEquations1D)
+  return ("η", "v", "D", "w", "η")
+end
+function varnames(::typeof(prim2cons), ::HyperbolicSerreGreenNaghdiEquations1D)
+  return ("h", "hv", "b", "hw", "hη")
+end
```

Comment:
@JoshuaLampert How would you like to handle the name clash of the total water height $$\eta = h + b$$ and the new variable coming from the hyperbolic relaxation approach $$\eta \approx h$$?

Reply:
I already feared this question reading your paper :smile:
I don't know what the best answer is, but maybe call it $H$ (I think we don't have $H$ right now) and clarify in docstrings what this is exactly?

Reply:
I have chosen `H`

Reply:
Thanks!